controller/src/controller/storage/docker.py

```
@@ -143,30 +145,84 @@ async def list_tags(self, config: DockerSourceOptions) -> list[str]:
-            raise DockerRegistryError.from_exception(e) from e
-        except Exception as e:
-            error = f"{type(e).__name__}: {e!s}"
-            msg = f"Cannot parse response from Docker registry: {error}"
-            raise DockerRegistryError(msg, method="GET", url=url) from e
-        else:
-            filtered = filter_arch_tags(tags)
-            self._logger.debug(
-                "Listed all image tags",
-                registry=config.registry,
-                repository=config.repository,
-                count=len(filtered),
-            )
-            return filtered
+                if r.status_code == 401:
+                    headers = await self._authenticate(config.registry, r)
+                    r = await self._client.get(url, headers=headers)
+                r.raise_for_status()
+                tags = r.json()["tags"]
+            except HTTPError as e:
+                raise DockerRegistryError.from_exception(e) from e
+            except Exception as e:
+                error = f"{type(e).__name__}: {e!s}"
+                msg = f"Cannot parse response from Docker registry: {error}"
+                raise DockerRegistryError(msg, method="GET", url=url) from e
+            else:
+                filtered = filter_arch_tags(tags)
+                count = len(filtered)
+                self._logger.debug(
+                    f"Listed {count} image tags",
+                    registry=config.registry,
+                    repository=config.repository,
+                    count=count,
+                )
+                current_tags = set(filtered)
+                duplicates = current_tags.intersection(all_filtered_tags)
+                if duplicates:
+                    tag_word = "tag" if len(duplicates) == 1 else "tags"
+                    self._logger.error(
+                        f"Duplicate {tag_word}: {duplicates}"
+                        f" Bailing out of tag-reading loop"
+                    )
```

I think this approach makes it likely that we'll never notice the problem, since we're just logging a message and not reporting an error anywhere, and we're then blindly continuing with a truncated list of tags.

I'm also a bit worried that this will be fragile. If someone adds a tag between the time we retrieve the first page of the tag list and the time we retrieve the second page, then depending on the pagination algorithm, we'll see the last tag of the first page a second time as it gets pushed down into the second tag. With this algorithm, we then abort reading further tags and proceed with a partial tag list, which isn't the end of the world but could result in advertising some weird images.

I think there's a limit to how much protection we can have against malicious or broken Docker servers, but if we're going to pick some failure mode, I think we should at least report an exception to Slack/Sentry so that there's some real noise, and I'd be more inclined to ensure that we never retrieve the same URL twice rather than worrying about duplicate tags, which I think might happen accidentally due to accidents of timing depending on the pagination algorithm used by the server.